transcendence-app/src/chat/chatroom/infrastructure/db/postgres/chatroom.postgres.repository.ts

```diff
@@ -60,47 +66,148 @@ export class ChatroomPostgresRepository
 
   async getPaginatedChatrooms(
     paginationDto: Required<PaginationWithSearchQueryDto>,
-  ): Promise<Chatroom[] | null> {
-    const { limit, offset, sort, search } = paginationDto;
-    const orderBy =
-      sort === BooleanString.True ? ChatroomKeys.NAME : ChatroomKeys.ID;
-    const chatroomsData = await makeQuery<Chatroom>(this.pool, {
-      text: `SELECT *
-             FROM ${this.table}
-             WHERE ${ChatroomKeys.NAME} ILIKE $1
-             ORDER BY ${orderBy}
+  ): Promise<GenericChat[] | null> {
+    const { limit, offset, search } = paginationDto;
+    const chatrooms = await makeQuery<GenericChatData>(this.pool, {
+      text: `WITH crMsg AS (SELECT m.${ChatroomMessageKeys.CHATROOM_ID},
+                                   m.${ChatroomMessageKeys.USER_ID},
+                                   m.${ChatroomMessageKeys.ID},
+                                   m.${ChatroomMessageKeys.CREATED_AT}
+                            FROM ${table.CHATROOM_MESSAGE} m),
+                  crDateProvider AS (SELECT m.${ChatroomMessageKeys.CHATROOM_ID},
+                                            ROW_NUMBER() OVER (
+                                              PARTITION BY m.${ChatroomMessageKeys.CHATROOM_ID} ORDER BY m.${ChatroomMessageKeys.CREATED_AT} DESC
+                                              ) AS "rowNumber",
+                                            m.${ChatroomMessageKeys.ID}
+                                     FROM crMsg m),
+                  crMsgIds AS (SELECT dp.${ChatroomMessageKeys.ID} AS "msgId"
+                               FROM crDateProvider dp
+                               WHERE dp."rowNumber" = 1),
+                  crMsgData AS (SELECT cm.${ChatroomMessageKeys.CHATROOM_ID},
+                                       u.${userKeys.USERNAME} AS "lastMsgSenderUsername",
+                                       cm.${ChatroomMessageKeys.CONTENT},
+                                       cm.${ChatroomMessageKeys.CREATED_AT}
+                                FROM ${table.CHATROOM_MESSAGE} cm
+                                       INNER JOIN crMsgIds mi ON cm.${ChatroomMessageKeys.ID} = mi."msgId"
+                                       LEFT JOIN ${table.USERS} u ON cm.${ChatroomMessageKeys.USER_ID} = u.${userKeys.ID}
+                                ORDER BY cm.${ChatroomMessageKeys.CREATED_AT})
+             SELECT cr.${ChatroomKeys.AVATAR_ID},
+                    cr.${ChatroomKeys.AVATAR_X},
+                    cr.${ChatroomKeys.AVATAR_Y},
+                    '/chatroom/' || cr.${ChatroomKeys.ID}            AS "url",
+                    cr.${ChatroomKeys.NAME},
+                    crmd."lastMsgSenderUsername",
+                    crmd.${ChatroomMessageKeys.CONTENT}::varchar(20) AS "lastMessage",
```

Comment:
varchar(20) for any particular reason?

Reply:
Because I wanted to truncate it to a sensible, small number of characters, and this seemed the simplest, and cleanest way to do it.

Reply:
elipsis is already implemented in the Row component. Either way, I've made it to return a text type. Should be good to go.